<!-- RGB-D Visual odometry -->

<launch>

  # ORB, SURF, GTF, KLT
  <arg name="detector_type" default="GFT"/> 

  # ICP, ICPModel, RANSAC
  <arg name="reg_type" default="ICP"/> 

  <node pkg="ccny_rgbd" type="visual_odometry_node" name="visual_odometry_node" 
    output="screen">

    <param name="fixed_frame" value="/odom"/>
    <param name="base_frame"  value="/camera_link"/>
       
    #### features #####################################
    
    #  ORB, SURF, or GFT (Good features to track)
    <param name="feature/detector_type"       value="$(arg detector_type)"/> 
    <param name="feature/compute_descriptors" value="false"/>
    <param name="feature/smooth"              value="0"/>
    <param name="feature/max_range"           value="7.0"/>
    <param name="feature/max_stdev"           value="0.03"/>
    <param name="feature/show_image"          value="true"/>
    <param name="feature/show_keypoints"      value="true"/>
    <param name="feature/publish_features"    value="true"/>
    <param name="feature/publish_covariances" value="false"/>

    #### features: GFT ################################

    <param name="feature/GFT/n_features"   value = "400"/>
    <param name="feature/GFT/min_distance" value = "0.0"/>
    <param name="feature/GFT/grid_cells"   value = "1"/>

    #### features: SURF ###############################
  
    <param name="feature/SURF/hessian_threshold" value = "600"/>
    <param name="feature/SURF/upright"           value = "false"/>

    #### features: ORB ###############################
  
    <param name="feature/ORB/n_features"     value = "300"/>
    <param name="feature/ORB/edge_threshold" value = "31"/>

    #### registration #################################

    <param name="reg/reg_type"          value="$(arg reg_type)"/>
    <param name="reg/motion_constraint" value="0"/>

    #### registration: ICP ############################

    <param name="reg/ICPProbModel/max_iterations" value="10"/>
    <param name="reg/ICP/history_size"            value="5"/>
    <param name="reg/ICP/max_corresp_dist_eucl"   value="0.15"/>
    <param name="reg/ICP/publish_model"           value="true"/>

    #### registration: ICP Prob Model #################

    <param name="reg/ICPProbModel/max_iterations"        value="10"/>
    <param name="reg/ICPProbModel/max_model_size"        value="3000"/>
    <param name="reg/ICPProbModel/n_nearest_neighbors"   value="4"/>
    <param name="reg/ICPProbModel/max_assoc_dist_mah"    value="10.0"/>
    <param name="reg/ICPProbModel/max_corresp_dist_eucl" value="0.15"/>
    <param name="reg/ICPProbModel/publish_model"         value="true"/>
    <param name="reg/ICPProbModel/publish_model_cov"     value="false"/>
  </node>

</launch>

